docs: improve and consolidate Realtime guide (#9359)
supabase/supabase: apps/reference/docs/guides/realtime/quickstart.mdx

---
id: quickstart
title: Quickstart
description: Getting started with Realtime's Features
---

Now that you're acquainted with Realtime's Broadcast, Presence, and Postgres CDC features, you can familiarize yourself with Realtime's API by reproducing the core features of Realtime's Multiplayer demo found at [multiplayer.dev](https://multiplayer.dev).

<div class="video-container">
  <iframe
    src="https://www.youtube-nocookie.com/embed/BelYEMJ2N00"
    frameBorder="1"
    allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
    allowFullScreen
  ></iframe>
</div>

## Prerequisites

- Install the `supabase-js` client:

```bash
npm install @supabase/supabase-js
```

> - Get the public `anon` access token from your project's [API settings](https://app.supabase.com/project/_/settings/api).

Author: @dannykng sounds good. Let me know what you think: https://github.com/supabase/supabase/pull/9359/commits/491dd38e3e0da348f0fa5cb467bd14dc96da6859.

## Broadcast

Broadcast allows a client to send messages and multiple clients to receive the messages.

The broadcasted messages are ephemeral. They are not persisted to the database and are directly relayed through the Realtime servers. This is ideal for sending information like cursor positions where minimal latency is important, but persisting them is not.

### Cursor Positions

For this example, cursor positions will be randomly generated.

> You can start sending a client's cursor positions to other clients in channel `room1`:

Author: @dannykng agreed. I tweaked it slightly to say that the demo uses actual cursor positions but this example has randomly generated ones: https://github.com/supabase/supabase/pull/9359/commits/5765beda0178f24663ed291d64fc8437c5cd2567.

// Create channels with the same name for both the broadcasting and receiving clients.
const channel = supabase.channel('room1')

// Subscribe registers your client with the server
channel.subscribe((status) => {
  if (status === 'SUBSCRIBED') {
    // now you can start broadcasting cursor positions
    setInterval(() => {
      channel.send({
        type: 'broadcast',
        event: 'cursor-pos',
        payload: { x: Math.random(), y: Math.random() },
      })
      console.log(status)
    }, 100)
  }
})
```

:::info
JavaScript client has a default rate limit of 1 Realtime event every 100 milliseconds that's configured by `eventsPerSecond`.
:::

Another client can subscribe to channel `room1` and receive cursor positions:

```js
// Supabase client setup

// Listen to broadcast messages.
supabase
  .channel('room1')
  .on('broadcast', { event: 'cursor-pos' }, (payload) => console.log(payload))
  .subscribe((status) => {
    if (status === 'SUBSCRIBED') {
      // your callback function will now be called with the messages broadcast by the other client
    }
  })
```

:::info
`type` must be `broadcast` and the `event` must match for clients subscribed to the channel.
:::

> ### Roundtrip Latency
> 
> You can configure the channel so that the server must return acknowledgement that it received the `broadcast` message. This is useful if you want to measure roundtrip latency:

Reviewer: ```suggestion
## Roundtrip Latency

You can also configure the channel so that the server must return an acknowledgement that it received the `broadcast` message. This is useful if you want to measure the roundtrip latency:
```

```js
// Supabase client setup

const channel = supabase.channel('calc-latency', {
  config: {
    broadcast: { ack: true }
  }
})

channel.subscribe(async (status) => {
    if (status === 'SUBSCRIBED) {
      const begin = performance.now()

      await channel.send({
        type: 'broadcast',
        event: 'latency',
        payload: {},
      })

      const end = performance.now()

      console.log(`Latency is ${end - begin} milliseconds`)
    }
  })
```

> ## Presence
> 
> Presence stores and synchronize shared state across clients.
> 
> The `sync` event is triggered whenever the shared state changes. The `join` event is triggered when new clients join the channel and `leave` event is triggered when clients leave.
> 
> Each client can use the channel's `track` method to store an object in shared state. Each client can only track one object, and if `track` is called again by the same client, then the new object will overwrite the previously tracked object in the shared state.
> 
> ### Online Users
> 
> You can use one client to track and display users who are online:

Reviewer: ```suggestion
## Track and display which users are online

[Presence](/docs/guides/realtime#presence) stores and synchronize shared state across clients. The `sync` event is triggered whenever the shared state changes. The `join` event is triggered when new clients join the channel and `leave` event is triggered when clients leave.

Each client can use the channel's `track` method to store an object in shared state. Each client can only track one object, and if `track` is called again by the same client, then the new object overwrites the previously tracked object in the shared state. You can use one client to track and display users who are online:
```

```js
// Supabase client setup

const channel = supabase.channel('online-users', {
  config: {
    presence: {
      key: 'user1',
    },
  },
})

channel.on('presence', { event: 'sync' }, () => {
  console.log('Online users: ', channel.presenceState())
})

channel.on('presence', { event: 'join' }, ({ newPresences }) => {
  console.log('New users have joined: ', newPresences)
})

channel.on('presence', { event: 'leave' }, ({ leftPresences }) => {
  console.log('Users have left: ', newPresences)
})

channel.subscribe(async (status) => {
  if (status === 'SUBSCRIBED') {
    const status = await channel.track({ online_at: new Date().toISOString() })
    console.log(status)
  }
})
```

Then you can use another client to add another user to the channel's Presence state:

```js
// Supabase client setup

const channel = supabase.channel('online-users', {
  config: {
    presence: {
      key: 'user2',
    },
  },
})

// Presence event handlers setup

channel.subscribe(async (status) => {
  if (status === 'SUBSCRIBED') {
    const status = await channel.track({ online_at: new Date().toISOString() })
    console.log(status)
  }
})
```

> :::info
> If presence key is not given then server will generate a random UUID. `type` must be `presence` and `event` must be either `sync`, `join`, or `leave`.
> :::

Reviewer: ```suggestion
If a channel is set up without a presence key, the server generates a random UUID. `type` must be `presence` and `event` must be either `sync`, `join`, or `leave`.
```

> ## Postgres CDC
> 
> Postgres Change Data Capture (CDC) enables your client to insert, update, or delete database records and have the changes sent to clients.
> 
> ### Insert and Receive Persisted Messages
> 
> Start by creating a `messages` table to keep track of messages created by users in specific rooms:

Reviewer: ```suggestion
## Insert and Receive Persisted Messages

[Postgres Change Data Capture (CDC)](/docs/guides/realtime#postgres-cdc) enables your client to insert, update, or delete database records and send the changes to clients. Create a `messages` table to keep track of messages created by users in specific rooms:
```

```sql
create table messages (
  id serial primary key,
  message text,
  user_id text,
  room_id text,
  created_at timestamptz default now()
)

alter table messages enable row level security;

create policy "anon_ins_policy"
ON messages
for insert
to anon
with check (true);

create policy "anon_sel_policy"
ON messages
for select
to anon
using (true);
```

> Make sure to create publication `supabase_realtime`, if it doesn't already exist, and add `messages` table to the publication:

Reviewer: ```suggestion
If it doesn't already exist, create a `supabase_realtime` publication and add `messages` table to the publication:
```
